src/community/graticule/src/main/java/org/geotools/data/graticule/GraticuleFeatureSource.java

```diff
@@ -9,58 +9,69 @@
 import java.util.List;
 import org.geotools.api.data.FeatureReader;
 import org.geotools.api.data.Query;
-import org.geotools.api.data.SimpleFeatureSource;
 import org.geotools.api.feature.simple.SimpleFeature;
 import org.geotools.api.feature.simple.SimpleFeatureType;
-import org.geotools.api.feature.type.Name;
+import org.geotools.api.filter.Filter;
+import org.geotools.data.graticule.gridsupport.LineFeatureBuilder;
 import org.geotools.data.store.ContentEntry;
 import org.geotools.data.store.ContentFeatureSource;
+import org.geotools.feature.simple.SimpleFeatureTypeBuilder;
 import org.geotools.geometry.jts.ReferencedEnvelope;
+import org.locationtech.jts.geom.LineString;
 
-public class GraticuleFeatureSource extends ContentFeatureSource implements SimpleFeatureSource {
+public class GraticuleFeatureSource extends ContentFeatureSource {
     private final List<Double> steps;
 
-    private final SimpleFeatureType schema;
     private final ReferencedEnvelope bounds;
-    private final GraticuleDataStore parent;
 
     public GraticuleFeatureSource(
-            ContentEntry entry,
-            Query query,
-            List<Double> steps,
-            ReferencedEnvelope bounds,
-            SimpleFeatureType schema) {
-        super(entry, query);
+            ContentEntry entry, List<Double> steps, ReferencedEnvelope bounds) {
+        super(entry, Query.ALL);
 
-        this.parent = (GraticuleDataStore) entry.getDataStore();
         this.steps = steps;
         this.bounds = bounds;
-        this.schema = schema;
+    }
+
+    List<Double> getSteps() {
+        return steps;
     }
 
     @Override
     protected ReferencedEnvelope getBoundsInternal(Query query) throws IOException {
+        if (query.getFilter() != Filter.INCLUDE) {
```

Comment:
are there not filters that should return bounds?

Reply:
According to the javadoc of FeatureSource:

```
@returns The bounding envelope of the feature data; or {@code null} if the bounds are unknown or too costly to calculate.
``` 

So the idea is to return bounds if it's fast, otherwise not. In this case, we are in a debatable situation... features are generated in memory so the computation should be fast unless there is a massive number of lines.. which some madman could indeed configure (and risk bringing GeoServer to OOM).

The implementation here takes the easy route. Did you ask me for everything? Configured bounds. Otherwise, if the user really wants it, you can use ``DataUtilities.bounds``, which will scan every single feature.